### <a name='0'></a>Import libraries

In [1]:
import numpy as np
import pandas as pd

np.__version__

'1.19.2'

### <a name='1'></a> Exercise 191
From the _data_ dictionary below, create a _DataFrame_ object and assign it to the _df_ variable.

```
data = {
    'size': ['XL', 'L', 'M', 'L', 'M'],
    'color': ['red', 'green', 'blue', 'green', 'red'],
    'gender': ['female', 'male', 'male', 'female', 'female'],
    'price': [199.0, 89.0, 99.0, 129.0, 79.0],
    'weight': [500, 450, 300, 380, 410],
    'bought': ['yes', 'no', 'yes', 'no', 'yes']
}
```



In [2]:
data = {
    'size': ['XL', 'L', 'M', 'L', 'M'],
    'color': ['red', 'green', 'blue', 'green', 'red'],
    'gender': ['female', 'male', 'male', 'female', 'female'],
    'price': [199.0, 89.0, 99.0, 129.0, 79.0],
    'weight': [500, 450, 300, 380, 410],
    'bought': ['yes', 'no', 'yes', 'no', 'yes']
}
df = pd.DataFrame(data=data)
df

size  color  gender  price  weight bought
0   XL    red  female  199.0     500    yes
1    L  green    male   89.0     450     no
2    M   blue    male   99.0     300    yes
3    L  green  female  129.0     380     no
4    M    red  female   79.0     410    yes

Display basic information about the _df_ object.

In [3]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   size    5 non-null      object 
 1   color   5 non-null      object 
 2   gender  5 non-null      object 
 3   price   5 non-null      float64
 4   weight  5 non-null      int64  
 5   bought  5 non-null      object 
dtypes: float64(1), int64(1), object(4)
memory usage: 368.0+ bytes


Convert the variables to individual types:
* size -> _category_
* color -> _category_
* gender -> _category_
* weight -> _float_
* bought -> _category_

In [4]:
col_list = df.select_dtypes(include = ['object']).columns
for col in col_list:
    df[col] = df[col].astype('category')
    
df['weight'] = df['weight'].astype('float')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   size    5 non-null      category
 1   color   5 non-null      category
 2   gender  5 non-null      category
 3   price   5 non-null      float64 
 4   weight  5 non-null      float64 
 5   bought  5 non-null      category
dtypes: category(4), float64(2)
memory usage: 628.0 bytes


### <a name='2'></a> Exercise 192
Using the _scikit-learn_ machine learning library and the _LabelEncoder_ class, encode 0-1 for the _bought_ variable. Permanently assign changes to the _df_ object and display this object.

In [5]:
df.bought

0    yes
1     no
2    yes
3     no
4    yes
Name: bought, dtype: category
Categories (2, object): ['no', 'yes']

In [6]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()
df.bought = le.fit_transform(df.bought)
df.bought

0    1
1    0
2    1
3    0
4    1
Name: bought, dtype: int32

Use the appropriate method of the _LabelEncoder_ class to restore the original form of the _bought_ variable. Note the type of the _bought_ variable.

In [7]:
df['bought'] = le.inverse_transform(df['bought'])
df['bought'] = df['bought'].astype('category')
df

size  color  gender  price  weight bought
0   XL    red  female  199.0   500.0    yes
1    L  green    male   89.0   450.0     no
2    M   blue    male   99.0   300.0    yes
3    L  green  female  129.0   380.0     no
4    M    red  female   79.0   410.0    yes

### <a name='3'></a> Exercise 193
Using the _scikit-learn_ machine learning library and the _OneHotEncoder_ class, encode 0-1 for the _size_ variable. Don't assign changes to the _df_ object.

Use the _sparse = False_ parameter of the _OneHotEncoder_ class.

In [8]:
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder(sparse = False)
enc.fit_transform(df[['size']])

array([[0., 0., 1.],
       [1., 0., 0.],
       [0., 1., 0.],
       [1., 0., 0.],
       [0., 1., 0.]])

Using the _scikit-learn_ machine learning library and the _OneHotEncoder_ class, encode 0-1 for the _size_ variable. Don't assign changes to the _df_ object.

Use the _sparse = False_ ** parameter and _drop = first _ ** of the _OneHotEncoder_ class.

In [9]:
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder(sparse = False,drop = 'first')
enc.fit_transform(df[['size']])

array([[0., 1.],
       [0., 0.],
       [1., 0.],
       [0., 0.],
       [1., 0.]])

### <a name='4'></a> Exercise 194
Using the _pandas_ library and the _get_dummies () _ function, present all categorical variables of the _df_ object in the form of 0-1 encoding. Do not permanently assign changes to the _df_ object.

In [10]:
pd.get_dummies(df)

price  weight  size_L  size_M  size_XL  color_blue  color_green  color_red  \
0  199.0   500.0       0       0        1           0            0          1   
1   89.0   450.0       1       0        0           0            1          0   
2   99.0   300.0       0       1        0           1            0          0   
3  129.0   380.0       1       0        0           0            1          0   
4   79.0   410.0       0       1        0           0            0          1   

   gender_female  gender_male  bought_no  bought_yes  
0              1            0          0           1  
1              0            1          1           0  
2              0            1          0           1  
3              1            0          1           0  
4              1            0          0           1

Using the _pandas_ library and the _get_dummies () _ function, present all categorical variables of the _df_ object in the form of 0-1 encoding. Use the _drop_first = True_ parameter. Do not permanently assign changes to the _df_ object.

In [11]:
pd.get_dummies(df,drop_first = True)

price  weight  size_M  size_XL  color_green  color_red  gender_male  \
0  199.0   500.0       0        1            0          1            0   
1   89.0   450.0       0        0            1          0            1   
2   99.0   300.0       1        0            0          0            1   
3  129.0   380.0       0        0            1          0            0   
4   79.0   410.0       1        0            0          1            0   

   bought_yes  
0           1  
1           0  
2           1  
3           0  
4           1

Using the _pandas_ library and the _get_dummies () _ function, present the ** variable _size _ ** of the _df_ object as 0-1 encoding. Use the _drop_first = True_ parameter. Do not permanently assign changes to the _df_ object.

In [12]:
pd.get_dummies(df['size'],drop_first = True)

M  XL
0  0   1
1  0   0
2  1   0
3  0   0
4  1   0

### <a name='5'></a> Exercise 195
Using the _scikit-learn_ machine learning library and the _StandardScaler_ class, scale the _weight_ variable. Don't assign changes to the _df_ object.

In [13]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
scaler.fit_transform(df[['weight']])

array([[ 1.3660019 ],
       [ 0.62360956],
       [-1.60356745],
       [-0.41573971],
       [ 0.02969569]])

### <a name='6'></a> Exercise 196
Re-load the dictionary _data_ into the _DataFrame_ object named _df_.

In [14]:
df = pd.DataFrame(data = data)
df

size  color  gender  price  weight bought
0   XL    red  female  199.0     500    yes
1    L  green    male   89.0     450     no
2    M   blue    male   99.0     300    yes
3    L  green  female  129.0     380     no
4    M    red  female   79.0     410    yes

Prepare data for the machine learning model:
* using the LabelEncoder class, code 0-1 for the _bought_ variable
* standardize the _weight_ and _price_ variables
* code 0-1 of the remaining categorical variables using the _get_dummies () function ._

Assign changes to the _df_prep_ variable.

In [15]:
df_prep = df.copy()

from sklearn import preprocessing
le = preprocessing.LabelEncoder()
df_prep['bought'] = le.fit_transform(df['bought'])

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
df[['weight','price']] = scaler.fit_transform(df[['weight','price']])

other = pd.get_dummies(df[['size','color','gender']])
df_prep = pd.concat([df_prep,other],axis=1)
df_prep = df_prep.drop(columns=['size','color','gender'])

df_prep

price  weight  bought  size_L  size_M  size_XL  color_blue  color_green  \
0  199.0     500       1       0       0        1           0            0   
1   89.0     450       0       1       0        0           0            1   
2   99.0     300       1       0       1        0           1            0   
3  129.0     380       0       1       0        0           0            1   
4   79.0     410       1       0       1        0           0            0   

   color_red  gender_female  gender_male  
0          1              1            0  
1          0              0            1  
2          0              0            1  
3          0              1            0  
4          1              1            0

### <a name='7'></a> Exercise 197
Copy the _df_prep_ object into the _data_ variable. Then take the _bought_ column out of that object and assign it to the _target_ variable.

In [16]:
data = df_prep.copy()
target = data.pop('bought')

Display the first 5 rows of the _data_ object. Make sure there is no _bought_ column in it.

In [17]:
data.head()

price  weight  size_L  size_M  size_XL  color_blue  color_green  color_red  \
0  199.0     500       0       0        1           0            0          1   
1   89.0     450       1       0        0           0            1          0   
2   99.0     300       0       1        0           1            0          0   
3  129.0     380       1       0        0           0            1          0   
4   79.0     410       0       1        0           0            0          1   

   gender_female  gender_male  
0              1            0  
1              0            1  
2              0            1  
3              1            0  
4              1            0

Display the first 5 lines of the _target_ object.

In [18]:
target.head()

0    1
1    0
2    1
3    0
4    1
Name: bought, dtype: int32

### <a name='8'></a> Exercise 198
Using the _scikit-learn_ library and the _train_test_split_ function, divide the data from the _data_ and _target_ objects into a training set (_X\_train_, _y\_train_) and a test set (_X\_test_, _y\_test_). Assign 20% of the samples to the test set.

View size of objects:
* _X\_train_
* _X\_test_
* _y\_train_
* _y\_test_

In [19]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    data, target, test_size=0.2)

print(f'X_trian shape: {X_train.shape}')
print(f'X_test shape: {X_test.shape}')
print(f'y_trian shape: {y_train.shape}')
print(f'y_test shape: {y_test.shape}')

X_trian shape: (4, 10)
X_test shape: (1, 10)
y_trian shape: (4,)
y_test shape: (1,)


### <a name='9'></a> Exercise 199
Combine the _X\_train_ and _X\_test_ objects into one _DataFrame_ object named _X_.

In [20]:
X = pd.concat([X_train,X_test],axis = 0)
X

price  weight  size_L  size_M  size_XL  color_blue  color_green  color_red  \
3  129.0     380       1       0        0           0            1          0   
1   89.0     450       1       0        0           0            1          0   
4   79.0     410       0       1        0           0            0          1   
0  199.0     500       0       0        1           0            0          1   
2   99.0     300       0       1        0           1            0          0   

   gender_female  gender_male  
3              1            0  
1              0            1  
4              1            0  
0              1            0  
2              0            1

### <a name='10'></a> Exercise 200
Combine the _y\_train_ and _y\_test_ objects into one _DataFrame_ object named _y_.

In [21]:
y = pd.concat([y_train,y_test],axis = 0)
y

3    0
1    0
4    1
0    1
2    1
Name: bought, dtype: int32

Combine the _X_ and _y_ objects into one _DataFrame_.

In [22]:
pd.concat([X,y],axis = 1)

price  weight  size_L  size_M  size_XL  color_blue  color_green  color_red  \
3  129.0     380       1       0        0           0            1          0   
1   89.0     450       1       0        0           0            1          0   
4   79.0     410       0       1        0           0            0          1   
0  199.0     500       0       0        1           0            0          1   
2   99.0     300       0       1        0           1            0          0   

   gender_female  gender_male  bought  
3              1            0       0  
1              0            1       0  
4              1            0       1  
0              1            0       1  
2              0            1       1